safenode/src/bin/kadnode.rs

```diff
@@ -1,233 +1,178 @@
+    // address.
+    // match opt.listen_adress {
+    //     Some(addr) => network_client
+    //         .start_listening(addr)
+    //         .await
+    //         .expect("Listening not to fail."),
+    //     None => network_client
+    //         .start_listening("/ip4/0.0.0.0/tcp/0".parse()?)
+    //         .await
+    //         .expect("Listening not to fail."),
+    // };
+
+    // In case the user provided an address of a peer on the CLI, dial it.
+    // if let Some(addr) = opt.peer {
+    //     let peer_id = match addr.iter().last() {
+    //         Some(Protocol::P2p(hash)) => PeerId::from_multihash(hash).expect("Valid hash."),
+    //         _ => return Err("Expect peer multiaddr to contain peer ID.".into()),
+    //     };
+    //     network_client
+    //         .dial(peer_id, addr)
+    //         .await
+    //         .expect("Dial to succeed");
+    // }
+    //
+
+    let mut api_clone = network_api.clone();
+    let storage_clone = storage.clone();
+    let (peer_dicovered_send, peer_dicovered_rx) = oneshot::channel();
+    let mut peer_dicovered_send = Some(peer_dicovered_send);
+    spawn(async move {
         loop {
-            select! {
-                cmd = receiver.recv().fuse() => {
-                    debug!("Cmd in: {cmd:?}");
-                    if let Some(SwarmCmd::Search(xor_name)) =  cmd {
-                        swarm.behaviour_mut().get_closest_peers_to_xorname(xor_name);
+            let event = match network_events.next().await {
+                Some(event) => event,
+                None => continue,
+            };
+            match event {
+                NetworkEvent::InboundSafeRequest { req, channel } => {
+                    // Reply with the content of the file on incoming requests.
+                    if let SafeRequest::GetChunk(xor_name) = req {
+                        let addr = ChunkAddress(xor_name);
+                        let chunk = storage_clone.query(&addr).await.unwrap();
+                        if let Err(err) = api_clone
+                            .send_safe_response(SafeResponse::Chunk(chunk), channel)
+                            .await
+                        {
+                            warn!("Error while sending safe_response: {err:?}");
+                        }
                     }
                 }
-
-                event = swarm.select_next_some() => match event {
-                    SwarmEvent::NewListenAddr { address, .. } => {
-                        info!("Listening in {address:?}");
-                    },
-                    SwarmEvent::Behaviour(SafeNetBehaviour::Mdns(mdns::Event::Discovered(list))) => {
-                        for (peer_id, multiaddr) in list {
-                            info!("Node discovered: {multiaddr:?}");
-                            swarm.behaviour_mut().kademlia.add_address(&peer_id, multiaddr);
+                NetworkEvent::PeerDiscoverd => {
```

Comment:
Isn't it possible to include the peer discovered in this enum? That way we could perhaps add it to the dht without a sleep further down? Or is that all code under the hood that we can't / shouldn't / don't want to tweak?

Reply:
`PeerDiscoverd` -> `PeerDiscovered`